### ISSUE TITLE
Keep qualified names tighter than the containing expression

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/dart_style#1557]
Here the first `js` is a library import prefix. It should not IMO be separated from the thing it qualifies.
(The formatting might be reasonable if the first `js` was a variable).
```dart
    return js.ExpressionStatement(
      js
          .js('# = #', [goto, js.number(label)])
          .withSourceInformation(sourceInformation),
    );
```
This would be better:
```dart
    return js.ExpressionStatement(
        js.js('# = #', [goto, js.number(label)])
        .withSourceInformation(sourceInformation),
    );
```

See https://dart-review.googlesource.com/c/sdk/+/383722/1/pkg/compiler/lib/src/js/rewrite_async.dart#315

**Comment by munificent** [comment]
I agree that the output here doesn't look great. That's often the case when you have a split call chain and the target (here just `js`) is shorter than 4 characters. It ends up looking like it's hanging out in space.

It's particularly egregious here because this isn't really a call chain with two method calls. It's a prefixed bare function call followed by a method call. Ideally, it would get formatted something like:

```dart
    return js.ExpressionStatement(
      js.js('# = #', [goto, js.number(label)])
          .withSourceInformation(sourceInformation),
    );
```

However, the formatter doesn't resolve anything. It doesn't know that `js` is a prefix and not an identifier expression.

You might argue that the formatter *could* know that because the prefixes are always easily textually found at the top of the file. However, that's not true when:

* We are formatting a part file and the imports are in some other file. (Enhanced part files will make this worse.)
* You are formatting a chunk of code which you will then paste into some other file.
* The prefix is shadowed by some other local declaration.

Sometimes I wish we used a different separator for prefixes, like `::`. But the formatter has to deal with what it's given and format with very little context.

There is *one* hacky heuristic the formatter uses to guess at what a name means. It uses capitalization to guess that an identifier is a type name versus a function name. That shows up in things like named constructors and static methods. If you have `Iterable.generate()`, the formatter won't split on that `.` because it sees that the first identifier is capitalized. Likewise, if you have `name.SomeClass()`, it assumes `name` is a prefix and not an identifier expression because of the subsequent capitalized name. Therefore it won't split after `name`.

However, that doesn't help here because the thing you're calling with a prefix is a top level function whose name is lowercase. :( If it was `js.Js`, then the formatter would give you:

```dart
    return js.ExpressionStatement(
      js.Js('# = #', [
        goto,
        js.number(label),
      ]).withSourceInformation(sourceInformation),
    );
```

But of course, you don't want to give a function a capitalized name.

So, I agree the output doesn't look great here, but it's probably about the best we can do. Sorry I don't have a great solution. :(

**Comment by Reprevise** [comment]
Not sure if this is the _correct_ issue but it looked similar. Here's this Flutter code:

Old:
```dart
Builder(
  builder: (context) => Column(
    spacing: 8,
    children: array.controls.map((e) {
      if (/* ... */) return /* ... */;

      return Row(
        children: [
          // ...
        ],
      );
    }).toList(growable: false),
  ),
),
```

New formatter:
```dart
Builder(
  builder:
      (context) => Column(
        spacing: 8,
        children: array.controls
            .map((e) {
              if (/* ... */) return /* ... */;

              return Row(
                children: [
                  // ...
                ],
              );
            })
            .toList(growable: false),
      ),
),
```

I think this is a pretty obvious downgrade. There's so much more indenting and the code is harder to read.

**Comment by munificent** [comment]
@Reprevise, your issue is a little different. It should be improved by the #1687 PR I recently landed. With that, your example gets formatted like:

```dart
Builder(
  builder:
      (context) => Column(
        spacing: 8,
        children: array.controls
            .map((e) {
              if (/* ... */) return /* ... */;

              return Row(
                children: [
                  // ...
                ],
              );
            })
            .toList(growable: false),
      ),
);
```

**Comment by Reprevise** [comment]
@munificent what's the difference?

**Comment by munificent** [comment]
Oh, heh. Oops, you're right in this case there is no difference. I thought it was running into #1466 where the formatter would give you:

```dart
Builder(
  builder:
      (context) => Column(
        spacing: 8,
        children:
            array.controls
                .map((e) {
                  if (/* ... */) return /* ... */;

                  return Row(
                    children: [
                      // ...
                    ],
                  );
                })
                .toList(growable: false),
      ),
);
```

But in this case, it looks like it at least wasn't *that* bad.

**Comment by munificent** [comment]
> I'm really not sure why we needed to change the line breaks and indentation of `.map().toList()` chain calls, it was perfect before.

The new tall style is a rewrite of the entire formatter so, for better or worse, it doesn't inherit the things the old formatter did right for free. The old formatter did a good job in some cases, but that same logic could break down and produce really nasty output in others.

For example, the old formatter would give you:

```dart
Builder(
  builder: (context) => Column(
    children: array.controls.map((e) {
      // ...
    }).toList(growable: false),
  ),
);
```

But if that property chain before `map()` was too long to fit, you would end up with:

```dart
Builder(
  builder: (context) => Column(
    children: array
        .controls.longerProperty
        .map((e) {
      // ...
    }).toList(growable: false),
  ),
);
```

Which is, I think, not great.

The new formatter avoids the latter, but isn't always as good at packing things in densely when they might fit.

Formatting method chains is surprisingly hard.
